cloudmarker/test/test_mongostore.py

```diff
@@ -0,0 +1,205 @@
+"""Tests for MongoDB store plugin."""
+
+
+import unittest
+from collections import defaultdict
+
+from pymongo import MongoClient
+
+from cloudmarker.stores import mongodbstore
+
+
+class DummyModel:
+    """A DummyModel to demonstrate constraint validation."""
+
+    __COLLECTION__ = "test_collection"
+    __ENFORCE__ = "error"  # ["error", "warn"]
+    __VALIDATOR__ = {
+        "$jsonSchema": {
+            "bsonType": "object",
+            "required": ["name"],
+            "properties": {
+                "name": {
+                    "bsonType": "string",
+                    "minLength": 3,
+                    "maxLength": 63,
+                    "description": "Name of the user, 2 < length < 64,required"
+                }
+            }
+        }
+    }
+
+    def __init__(self, name="", attr=defaultdict()):
+        """Initialize the DummyModel object for test."""
+        self.name = name
+        self.attr = attr
+
+    @staticmethod
+    def validator():
+        return DummyModel.__VALIDATOR__
+
+    @staticmethod
+    def enforce():
+        return DummyModel.__ENFORCE__
+
+    @staticmethod
+    def collection():
+        return DummyModel.__COLLECTION__
+
+    def marshal(self):
+        mp = {
+            "name": self.name
+        }
+
+        for k, v in self.attr.items():
+            mp[k] = v
+
+        mp["record_type"] = DummyModel.collection()
+
+        return mp
+
+
+class MongoDBStoreTest(unittest.TestCase):
+    """Tests for MongoDB store plugin.
+
+    All tests would require a running instance of mongoDB. You may use
+    the docker compose file in ``cloudmarker/test/data/docker-comppose.yml``
+    $ docker-compute -f cloudmarker/test/data/docker-comppose.yml up
+    """
+
+    def test_connection(self):
+        client = MongoClient(host='localhost', port=27017,
+                             username='dummy', password='password')
+
+        db = client["test"]
+
+        for collection in db.list_collection_names():
+            db[collection].drop()
+
+        store = mongodbstore.MongoDBStore('test', 'dummy',
```

Comment:
Use keywords argument for more clarity.

Reply:
@prateeknischal & @susam I have a curiosity about this test case. Why do we have an integration test case for this store plugin as currently, we don't have any CD in the pipeline? I am not pretty much sure if this integration test would serve any purpose and if it does what would it be. I was thinking if we can have a mock unit test to test the MongoDB store connect would fit in this scenario. Also is it that we are trying to test the pymongo API here? 
I would like to know your thoughts on this @susam.

Reply:
In general, it is good to have integration tests. As the project becomes more and more complex, integration tests would tell us if end-to-end functionality is running fine or not. At the same time, I think it is too soon to be writing integration tests right now.

Developing integration tests pose a many challenges: For example, how do we connect to a test instance of GCP only to perform integration tests? Where would we get such a test instance from? All these details have to be worked out. Integration tests also increases the build time, so we have to figure out a way to run integration tests less frequently only on certain builds (like on important milestones or before releases). Integration tests also add _a lot_ of complexity to the tests. Do we as a team enough free time to maintain these integration tests?

For stores based on open source components, it is a little easier to do integration tests. We can spin up Docker containers for MongoDB, Elasticsearc, etc. and test against these containers. But concerns about increased build times and maintenance overhead remain.

As far as unit testing with mock is concerned, you are right. That is the right way to do unit testing. I will spend sometime on exploring if we can use `mock.patch` to unit-test the stores without having to spin up actual instances of MongoDB, Elasticsearch, etc.

Our first priority is unit testing and better code coverage using unit tests. Integration tests is low priority right now. Since integration tests would be done less frequently, integration tests also won't count towards the code coverage metrics that are run on every build.